File: src/main/python/refinery_roborio_mcp/json_rpc.py
"""JSON-RPC 2.0 response builder for MCP protocol."""

import json

JSONRPC_VERSION = "2.0"


def result_response(id, result):
    """Build a JSON-RPC 2.0 success response."""
    return json.dumps({
        "jsonrpc": JSONRPC_VERSION,
        "id": id,
        "result": result,
    })


def error_response(id, code, message):
    """Build a JSON-RPC 2.0 error response."""
    return json.dumps({
        "jsonrpc": JSONRPC_VERSION,
        "id": id,
        "error": {"code": code, "message": message},
    })

File: src/main/cpp/JsonRpc.cpp
// Copyright (c) BioNanomics. All rights reserved.
// Licensed under the MIT License.

#include "refinery/mcp/JsonRpc.h"

#include <string>
#include <string_view>

namespace refinery::mcp {

std::string JsonRpc::ResultResponse(const wpi::json& id,
                                    const wpi::json& result) {
  wpi::json response = {
      {"jsonrpc", "2.0"}, {"id", id}, {"result", result}};
  return response.dump();
}

std::string JsonRpc::ErrorResponse(const wpi::json& id, int code,
                                   std::string_view message) {
  wpi::json response = {{"jsonrpc", "2.0"},
                         {"id", id},
                         {"error", {{"code", code}, {"message", std::string(message)}}}};
  return response.dump();
}

}  // namespace refinery::mcp
